src/app/static/src/app/components/plots/bubble/utils.ts

```diff
@@ -0,0 +1,78 @@
+import {BubbleIndicatorValuesDict, Dict, Filter, IndicatorValuesDict} from "../../../types";
+import {getColor} from "../../../store/filteredData/utils";
+import {Feature} from "geojson";
+import {ChoroplethIndicatorMetadata} from "../../../generated";
+
+export const toIndicatorNameLookup = (array: ChoroplethIndicatorMetadata[]) =>
+    array.reduce((obj, current) => {
+        obj[current.indicator] = current.name;
+        return obj
+    }, {} as Dict<string>);
+
+export const getFeatureIndicators = function (data: any[],
+                                              selectedFeatures: Feature[],
+                                              indicatorsMeta: ChoroplethIndicatorMetadata[],
+                                              minRadius: number,
+                                              maxRadius: number): Dict<BubbleIndicatorValuesDict> {
+
+    const result = {} as Dict<BubbleIndicatorValuesDict>;
+
+    const selectedAreaIds = selectedFeatures.map(f => f.properties!!.area_id);
+
+    const minArea = Math.PI * Math.pow(minRadius, 2);
+    const maxArea = Math.PI * Math.pow(maxRadius, 2);
+
+
+    for (const row of data) {
+        //TODO: exclude rows based on filters
+        //if (excludeRow(row, selectedRegionFilters)) {
+        //    continue;
+        //}
+
+        const areaId: string = row.area_id;
+
+        if (!selectedAreaIds.includes(areaId)) {
+            continue;
+        }
+
+        for (const metadata of indicatorsMeta) {
+
+            const indicator = metadata.indicator;
+
+            if (metadata.indicator_column && metadata.indicator_value != row[metadata.indicator_column]) {
+                //This data is in long format, and the indicator column's value does not match that for this indicator
+                continue;
+            }
+
+            if (!row[metadata.value_column]) {
+                //No value for this indicator in this row
+                continue;
+            }
+
+            const value = row[metadata.value_column];
+
+            if (!result[areaId]) {
+                result[areaId] = {} as BubbleIndicatorValuesDict;
+            }
+
+            const regionValues = result[areaId];
+            regionValues[indicator] = {
+                value: value,
+                color: getColor(value, metadata), //TODO: put this function. shared with Choropleth, in a more generic place
+                radius: getRadius(value, metadata.min, metadata.max, minArea, maxArea)
+            }
+        }
+    }
+
+    return result;
+};
+
+export const getRadius = function(value: number, minValue: number, maxValue: number, minArea: number, maxArea: number){
+    value = Math.min(maxValue, value);
```

Comment:
I don't think we even bother to check in the choropleth, we just assume the color function will be able to deal with it.. Throwing an error seems a bit  hard on the user - might just be badly designed metadata. Console log a warning?

Reply:
Although, actually, if we're not going to be calculating min and max values ourselves this problem goes away.

Reply:
warn and omit the value seems good :+1:

Reply:
Don't think we even need to do that now, since we're determining the actual min and max values ourselves, so unless there's a bug you shouldn't ever see a value outside the range.

Reply:
Nice